### ISSUE TITLE
Custom Metrics resources are empty

### ISSUE THREAD

**Issue by VinayVanama** [banzaicloud/banzai-charts#583]
Hi, I have installed prometheus-adapter helm chart on my EKS, deployment was successful but when I'm trying to query for 
kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1 
I'm getting resources as [] empty.
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"custom.metrics.k8s.io/v1beta1","resources":[]}

Can you please let me know what was the issue ? Where I'm wrong ?
Thanks

**Comment by sancyx** [comment]
Hi, first thing would be to check the logs of prom-adapter. You should see lots of metrics fetched from prometheus, or some error if something went wrong. Also make sure metrics-server is installed because it depends on EKS version, whether is installed by default.

**Comment by VinayVanama** [comment]
@sancyx my EKS version is v1.10.11, Even metrics-server is installed ! Below are the logs from my prom-adapter 

I1227 10:03:39.553077       1 round_trippers.go:417] curl -k -v -XPOST  -H "Accept: application/json, */*" -H "Content-Type: application/json" -H "User-Agent: adapter/v0.0.0 (linux/amd64) kubernetes/$Format" -H "Authorization: Bearer 
--------------
https://172.20.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews
I1227 10:03:39.559803       1 round_trippers.go:436] POST https://172.20.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews 201 Created in 6 milliseconds
I1227 10:03:39.559817       1 round_trippers.go:442] Response Headers:
I1227 10:03:39.559822       1 round_trippers.go:445]     Audit-Id: e5876b39-cf0b-42da-95f3-a19f493c5fae
I1227 10:03:39.559826       1 round_trippers.go:445]     Content-Type: application/json
I1227 10:03:39.559853       1 round_trippers.go:445]     Content-Length: 260
I1227 10:03:39.559859       1 round_trippers.go:445]     Date: Thu, 27 Dec 2018 10:03:39 GMT
I1227 10:03:39.559884       1 request.go:836] Response Body: {"kind":"SubjectAccessReview","apiVersion":"authorization.k8s.io/v1beta1","metadata":{"creationTimestamp":null},"spec":{"nonResourceAttributes":{"path":"/","verb":"get"},"user":"system:anonymous","group":["system:unauthenticated"]},"status":{"allowed":false}}
I1227 10:03:39.560062       1 authorization.go:59] Forbidden: "/", Reason: ""
I1227 10:03:39.560161       1 wrap.go:42] GET /: (7.291787ms) 403 [[Go-http-client/2.0] 10.0.9.233:47616]

**Comment by sancyx** [comment]
Yes there's are two parameters for prometheus-adapter chart: prometheus.url & prometheus.port, you can pass these like: helm install prometheus-adaptyer --set  prometheus.url='xxx' --set prometheus.port=xxx.

**Comment by VinayVanama** [comment]
> Yes there's are two parameters for prometheus-adapter chart: prometheus.url & prometheus.port, you can pass these like: helm install prometheus-adaptyer --set prometheus.url='xxx' --set prometheus.port=xxx.

yes I have done in the same way !

**Comment by sancyx** [comment]
Then it seems configuration is right, there's a connectivity issue to Prometheus server.

**Comment by VinayVanama** [comment]
@sancyx I have made it work by creating it from scratch ! Finally my config is working and I have deployed a sample-app and its showing metrics ! But when I deployed my application I'm unable to retrieve metrics like (http_requests) Do we need to write our own custom-metrics ? If yes ? Where should I write them and how can I use them to scale my pods ?

**Comment by sancyx** [comment]
Yes you had to create your custom metrics in your application and expose them to prometheus. This adapter makes promethues metrics visible to metrics-server. Once you can retrieve your custom metric with 'kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1' you can create a HorizontalPodAutoscaler in k8s which will take care of scaling you deployment. Checkout the example here: https://banzaicloud.com/blog/k8s-horizontal-pod-autoscaler/

**Comment by VinayVanama** [comment]
Hi @sancyx I'm unable to find where to write the custom rules ! Even in the above example you mentioned I'm not able to find them. Can you please guide me where should I write these ? Should I write these in my deployment.yaml or configmap.yaml or in the application code ?

**Comment by sancyx** [comment]
Hi @VinayVanama, you had to create a HorizontalPodAutoscaler object, where you can specify scale target, min/max replicas and metrics:

apiVersion: autoscaling/v2beta1
kind: HorizontalPodAutoscaler
metadata:
  name: wildfly-example
spec:
  scaleTargetRef:
    apiVersion: extensions/v1beta1
    kind: Deployment
    name: wildfly-example
  minReplicas: 1
  maxReplicas: 5
  metrics:
  - type: Resource
    resource:
      name: cpu
      targetAverageUtilization: 80
  - type: Resource
    resource:
      name: memory
      targetAverageValue: 1000Mi

**Comment by SankhaJay** [comment]
make sure the prometheus adapter is in the same namesapce as the other resorces. Change the 
 args:
        - --prometheus-url=http://prometheus-operated.<namespace_nmae>.svc:<port_name>/

re apply it and delete the current pod and deployment will ensure that new pod is automatically created.
